Doc/using/introspection.rst

```diff
@@ -0,0 +1,153 @@
+a general guide on how to approach this task, but users should still refer to
+the documentation of their installation for more accurate information regarding
+implementation details and availability.
+
+
+Introduction
+============
+
+Generally, the most portable way to do this is by running some Python code to
+extract information from the target interpreter, though this can often be
+automated with the use of built-in helpers (see
+:ref:`introspection-stdlib-clis`, :ref:`introspection-python-config-script`).
+
+However, when introspecting a Python installation, running code is often
+undesirable or impossible. For this purpose, Python defines a standard
+:ref:`static install description file <introspection-install-details>` (please
+see the :ref:`introspection-install-details-format` section for the format
+specification), which may be provided together with your interpreter (please
+see the :ref:`introspection-install-details-location` section for more details).
+When available, this file can be used to lookup certain details from the
+installation in question without having to run the interpreter.
+
+
+Reference
+=========
+
+
+.. _introspection-install-details:
+
+Install details file
+--------------------
+
+
+.. _introspection-install-details-format:
+
+Format
+~~~~~~
+
+The details file is JSON file and the top-level object is a dictionary
+containing the data described in :ref:`introspection-install-details-fields`.
+
+
+.. _introspection-install-details-fields:
+
+Fields
+~~~~~~
+
+This section documents some of the data fields that may be present. It is not
+extensive — different implementations, platforms etc. may include additional
+fields.
+
+Dictionaries represent sub-sections, and their names are separated with
+dots (``.``) in this documentation.
+
+``python.version``
+++++++++++++++++++
+
+:Type: ``str``
+:Description: A string representation of the Python version.
+
+``python.version_parts``
+++++++++++++++++++++++++
+
+:Type: ``dict[str, int | str]``
+:Description: A dictionary containing the different components of the Python
+              version, as found in :py:data:`sys.version_info`.
+
+``python.executable``
++++++++++++++++++++++
+
+:Type: ``str``
+:Description: An absolute path pointing to an interpreter from the installation.
+:Availability: May be omitted in situations where is is unavailable, there isn't
+               a reliable path that can be specified, etc.
+
+``python.stdlib``
++++++++++++++++++++++
+
+:Type: ``str``
+:Description: An absolute path pointing to the directory where the standard
+              library is installed.
+:Availability: May be omitted in situations where is is unavailable, there isn't
+               a reliable path that can be specified, etc.
+
+
+.. _introspection-install-details-location:
+
+Location
+~~~~~~~~
+
+On standard CPython installations, when available, the details file will be
+installed in the same directory as the standard library, with the name
```

Comment:
> In some cases the standard library is split. Can we name a specific file it will be alongside? (e.g. the tests below expect it to be next to `sysconfig.py`)

Right, I will clarify it is the pure part of the standard library, which corresponds to the sysconfig `stdlib` path.

> Also, on Windows, this implies that everyone should hard-code `Lib` into their paths, because the only paths you can resolve without running Python are `sys.prefix` and `sys.executable`. (I'm not sure how best to go about locating the stdlib on other platforms either, to be honest, but I trust you for those.)

I don't see a great solution for this, given that `sys.prefix` may contain multiple installations. My main two target use-cases are cross-compilation, where the user should be able to locate `install-details.json` themselves, and Python launchers, where, knowing that `stdlib` is within `sys.prefix`, should be able to search for the `**/install-details.txt` pattern in order to discover the available installation within a prefix.

The approach for Python launchers is not optimal, but it can be fixed later. Currently, I am just focusing on getting the file format down, and making the file available somewhere. Afterwards, we can look at the discoverability issue and consider for eg. having a global registry where these files could be additionally installed (eg. on Linux this could be the `~/.config/python/installations` and `/etc/python/installations` paths), but that's a whole other issue, and I am trying to break things down into smaller chunks so that I can actually hope to make meaningful progress.

What do you think?